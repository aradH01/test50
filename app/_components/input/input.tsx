"use client";

import { InputHTMLAttributes, forwardRef, ReactNode } from "react";
import { cn } from "@lib/utils";

export type InputVariant = "default" | "error" | "success" | "ghost";
export type InputSize = "default" | "sm" | "lg";

const getVariantClasses = (variant: InputVariant): string => {
  const variants = {
    default: "border-gray-300 focus:border-blue-500 focus:ring-blue-500",
    error:
      "border-red-500 focus:border-red-500 focus:ring-red-500 text-red-900 placeholder-red-300",
    success:
      "border-green-500 focus:border-green-500 focus:ring-green-500 text-green-900",
    ghost:
      "border-transparent focus:border-blue-500 focus:ring-blue-500 bg-gray-50 hover:bg-gray-100",
  };
  return variants[variant];
};

const getSizeClasses = (size: InputSize): string => {
  const sizes = {
    default: "h-10 px-3 py-2 text-sm",
    sm: "h-8 px-2 py-1 text-xs",
    lg: "h-12 px-4 py-3 text-base",
  };
  return sizes[size];
};

export type InputProps = Omit<InputHTMLAttributes<HTMLInputElement>, "size"> & {
  variant?: InputVariant;
  size?: InputSize;
  label?: string;
  error?: string;
  success?: string;
  helpText?: string;
  leftIcon?: ReactNode;
  rightIcon?: ReactNode;
};

export const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      className,
      variant = "default",
      size = "default",
      label,
      error,
      success,
      helpText,
      leftIcon,
      rightIcon,
      disabled,
      id,
      ...props
    },
    ref,
  ) => {
    const inputId = id || label?.toLowerCase().replace(/\s+/g, "-");
    const hasError = error || variant === "error";
    const hasSuccess = success || variant === "success";

    // Override variant if error or success props are provided
    const finalVariant = hasError ? "error" : hasSuccess ? "success" : variant;

    const baseClasses =
      "block w-full rounded-md border bg-white shadow-sm transition-colors focus:outline-none focus:ring-1 disabled:cursor-not-allowed disabled:bg-gray-50 disabled:text-gray-500";

    const inputContent = (
      <div className="relative">
        {leftIcon && (
          <div className="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none text-gray-400">
            {leftIcon}
          </div>
        )}
        <input
          ref={ref}
          id={inputId}
          className={cn(
            baseClasses,
            getVariantClasses(finalVariant),
            getSizeClasses(size),
            leftIcon && "pl-10",
            rightIcon && "pr-10",
            className,
          )}
          disabled={disabled}
          aria-describedby={
            error
              ? `${inputId}-error`
              : success
                ? `${inputId}-success`
                : helpText
                  ? `${inputId}-help`
                  : undefined
          }
          aria-invalid={hasError ? "true" : "false"}
          {...props}
        />
        {rightIcon && (
          <div className="absolute inset-y-0 right-0 flex items-center pr-3 pointer-events-none text-gray-400">
            {rightIcon}
          </div>
        )}
      </div>
    );

    if (label || error || success || helpText) {
      return (
        <div className="space-y-1">
          {label && (
            <label
              htmlFor={inputId}
              className={cn(
                "block text-sm font-medium",
                hasError
                  ? "text-red-700"
                  : hasSuccess
                    ? "text-green-700"
                    : "text-gray-700",
                disabled && "text-gray-400",
              )}
            >
              {label}
            </label>
          )}
          {inputContent}
          {error && (
            <p id={`${inputId}-error`} className="text-sm text-red-600">
              {error}
            </p>
          )}
          {success && !error && (
            <p id={`${inputId}-success`} className="text-sm text-green-600">
              {success}
            </p>
          )}
          {helpText && !error && !success && (
            <p id={`${inputId}-help`} className="text-sm text-gray-500">
              {helpText}
            </p>
          )}
        </div>
      );
    }

    return inputContent;
  },
);

Input.displayName = "Input";
