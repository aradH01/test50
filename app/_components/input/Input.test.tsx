/// <reference types="jest" />
/// <reference types="@testing-library/jest-dom" />

import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Input } from "./input";

// Helper component for testing with icons
const TestIcon = () => <span data-testid="test-icon">Icon</span>;

describe("Input", () => {
  // Arrange, Act, Assert pattern

  describe("Basic functionality", () => {
    it("renders input with placeholder", () => {
      // Arrange & Act
      render(<Input placeholder="Enter text" />);

      // Assert
      expect(screen.getByPlaceholderText("Enter text")).toBeInTheDocument();
    });

    it("accepts user input", async () => {
      // Arrange
      const user = userEvent.setup();
      render(<Input placeholder="Enter text" />);
      const input = screen.getByPlaceholderText("Enter text");

      // Act
      await user.type(input, "Hello World");

      // Assert
      expect(input).toHaveValue("Hello World");
    });

    it("forwards ref correctly", () => {
      // Arrange
      const ref = { current: null };

      // Act
      render(<Input ref={ref} />);

      // Assert
      expect(ref.current).toBeInstanceOf(HTMLInputElement);
    });
  });

  describe("Variants", () => {
    it("applies default variant classes", () => {
      // Arrange & Act
      render(<Input data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("border-gray-300");
    });

    it("applies error variant classes", () => {
      // Arrange & Act
      render(<Input variant="error" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("border-red-500");
    });

    it("applies success variant classes", () => {
      // Arrange & Act
      render(<Input variant="success" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("border-green-500");
    });

    it("applies ghost variant classes", () => {
      // Arrange & Act
      render(<Input variant="ghost" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("border-transparent");
    });
  });

  describe("Sizes", () => {
    it("applies default size classes", () => {
      // Arrange & Act
      render(<Input data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("h-10");
    });

    it("applies small size classes", () => {
      // Arrange & Act
      render(<Input size="sm" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("h-8");
    });

    it("applies large size classes", () => {
      // Arrange & Act
      render(<Input size="lg" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("h-12");
    });
  });

  describe("Label and messages", () => {
    it("renders with label", () => {
      // Arrange & Act
      render(<Input label="Email Address" />);

      // Assert
      expect(screen.getByLabelText("Email Address")).toBeInTheDocument();
      expect(screen.getByText("Email Address")).toBeInTheDocument();
    });

    it("associates label with input correctly", () => {
      // Arrange & Act
      render(<Input label="Email Address" />);
      const label = screen.getByText("Email Address");
      const input = screen.getByLabelText("Email Address");

      // Assert
      expect(label).toHaveAttribute("for", input.id);
    });

    it("displays error message", () => {
      // Arrange & Act
      render(<Input label="Password" error="Password is required" />);

      // Assert
      expect(screen.getByText("Password is required")).toBeInTheDocument();
      expect(screen.getByLabelText("Password")).toHaveAttribute(
        "aria-invalid",
        "true",
      );
    });

    it("displays success message", () => {
      // Arrange & Act
      render(<Input label="Username" success="Username is available" />);

      // Assert
      expect(screen.getByText("Username is available")).toBeInTheDocument();
    });

    it("displays help text", () => {
      // Arrange & Act
      render(<Input label="Email" helpText="We'll never share your email" />);

      // Assert
      expect(
        screen.getByText("We'll never share your email"),
      ).toBeInTheDocument();
    });

    it("prioritizes error over success message", () => {
      // Arrange & Act
      render(
        <Input label="Test" error="Error message" success="Success message" />,
      );

      // Assert
      expect(screen.getByText("Error message")).toBeInTheDocument();
      expect(screen.queryByText("Success message")).not.toBeInTheDocument();
    });
  });

  describe("Icons", () => {
    it("renders left icon", () => {
      // Arrange & Act
      render(<Input leftIcon={<TestIcon />} />);

      // Assert
      expect(screen.getByTestId("test-icon")).toBeInTheDocument();
    });

    it("renders right icon", () => {
      // Arrange & Act
      render(<Input rightIcon={<TestIcon />} />);

      // Assert
      expect(screen.getByTestId("test-icon")).toBeInTheDocument();
    });

    it("applies correct padding when left icon is present", () => {
      // Arrange & Act
      render(<Input leftIcon={<TestIcon />} data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("pl-10");
    });

    it("applies correct padding when right icon is present", () => {
      // Arrange & Act
      render(<Input rightIcon={<TestIcon />} data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveClass("pr-10");
    });
  });

  describe("Accessibility", () => {
    it("sets aria-invalid to false by default", () => {
      // Arrange & Act
      render(<Input data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveAttribute(
        "aria-invalid",
        "false",
      );
    });

    it("sets aria-invalid to true when error is present", () => {
      // Arrange & Act
      render(<Input error="Error message" data-testid="input" />);

      // Assert
      expect(screen.getByTestId("input")).toHaveAttribute(
        "aria-invalid",
        "true",
      );
    });

    it("links input to error message with aria-describedby", () => {
      // Arrange & Act
      render(<Input label="Test" error="Error message" />);
      const input = screen.getByLabelText("Test");

      // Assert
      expect(input).toHaveAttribute(
        "aria-describedby",
        expect.stringContaining("error"),
      );
    });
  });

  describe("Disabled state", () => {
    it("applies disabled styling and behavior", () => {
      // Arrange & Act
      render(<Input disabled data-testid="input" />);
      const input = screen.getByTestId("input");

      // Assert
      expect(input).toBeDisabled();
      expect(input).toHaveClass("disabled:cursor-not-allowed");
    });
  });

  // Snapshot test
  it("matches snapshot", () => {
    // Arrange & Act
    const { container } = render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        helpText="This is help text"
      />,
    );

    // Assert
    expect(container.firstChild).toMatchSnapshot();
  });
});
