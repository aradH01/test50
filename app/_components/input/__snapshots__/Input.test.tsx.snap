// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Input matches snapshot 1`] = `
<div
  class="space-y-1"
>
  <label
    class="block text-sm font-medium text-gray-700"
    for="test-input"
  >
    Test Input
  </label>
  <div
    class="relative"
  >
    <input
      aria-describedby="test-input-help"
      aria-invalid="false"
      class="block w-full rounded-md border bg-white shadow-sm transition-colors focus:outline-none focus:ring-1 disabled:cursor-not-allowed disabled:bg-gray-50 disabled:text-gray-500 border-gray-300 focus:border-blue-500 focus:ring-blue-500 h-10 px-3 py-2 text-sm"
      id="test-input"
      placeholder="Enter text"
    />
  </div>
  <p
    class="text-sm text-gray-500"
    id="test-input-help"
  >
    This is help text
  </p>
</div>
`;
